client/src/main/java/com/vesoft/nebula/storage/client/StorageClientImpl.java

```diff
@@ -276,71 +250,116 @@ public boolean put(int part, Map<String, String> values) {
+        if (!optional.isPresent()) {
+            return Optional.empty();
+        }
+        StorageService.Client client = optional.get();
+
         GetRequest request = new GetRequest();
         request.setSpace_id(space);
         Map<Integer, List<String>> parts = Maps.newHashMap();
         parts.put(part, Arrays.asList(key));
         request.setParts(parts);
         LOGGER.debug(String.format("Get Request: %s", request.toString()));
 
-        GeneralResponse response;
-        int retry = connectionRetry;
-        while (retry-- != 0) {
-            try {
-                response = client.get(request);
-                if (!isSuccess(response)) {
-                    for (ResultCode code : response.result.getFailed_codes()) {
-                        if (code.getCode() == ErrorCode.E_LEADER_CHANGED) {
-                            HostAddr addr = code.getLeader();
-                            if (addr != null && addr.getIp() != 0 && addr.getPort() != 0) {
-                                updateLeader(space, code.getPart_id(),
-                                        new HostAddr(addr.getIp(), addr.getPort()));
-                                connect(addr);
-                            }
-                        }
-                    }
-                } else {
-                    if (!leaders.get(space).containsKey(part)
-                            || leaders.get(space).get(part) != currentLeaderAddress) {
-                        updateLeader(space, part, currentLeaderAddress);
-                    }
-                    if (response.values.containsKey(key)) {
-                        return Optional.of(response.values.get(key));
-                    } else {
-                        return Optional.empty();
-                    }
-                }
-            } catch (TException e) {
-                LOGGER.error(String.format("Get Failed: %s", e.getMessage()));
-            }
+        Optional<Map<String, String>> result = doGet(space, client, request);
+        if (!result.isPresent() || !result.get().containsKey(key)) {
+            return Optional.empty();
+        } else {
+            return Optional.of(result.get().get(key));
         }
-        return Optional.empty();
     }
 
     /**
      * Get multi keys from part
      *
-     * @param part partitionID
-     * @param keys nebula keys
+     * @param space nebula space id
+     * @param keys  nebula keys
      * @return
      */
     @Override
-    public Optional<Map<String, String>> get(int part, List<String> keys) {
-        checkLeader(part);
-        GetRequest request = new GetRequest();
-        Map<Integer, List<String>> parts = Maps.newHashMap();
-        parts.put(part, keys);
-        request.setSpace_id(space);
-        request.setParts(parts);
-        LOGGER.debug(String.format("Get Request: %s", request.toString()));
+    public Optional<Map<String, String>> get(int space, List<String> keys) {
```

Comment:
Do we know space ID ? Maybe we need something to transform SpaceName into Space ID

Reply:
Ageed, we need to support in meta client later

Reply:
^_^